crates/miden-objects/src/block/nullifier_tree.rs

```diff
@@ -0,0 +1,297 @@
+        entries: impl IntoIterator<Item = (Nullifier, BlockNumber)>,
+    ) -> Result<Self, NullifierTreeError> {
+        let leaves = entries.into_iter().map(|(nullifier, block_num)| {
+            (nullifier.inner(), Self::block_num_to_leaf_value(block_num))
+        });
+
+        let smt = Smt::with_entries(leaves)
+            .map_err(NullifierTreeError::DuplicateNullifierBlockNumbers)?;
+
+        Ok(Self { smt })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the root of the nullifier SMT.
+    pub fn root(&self) -> RpoDigest {
+        self.smt.root()
+    }
+
+    /// Returns the number of spent nullifiers in this tree.
+    pub fn num_nullifiers(&self) -> usize {
+        self.smt.num_entries()
+    }
+
+    /// Returns a [`NullifierWitness`] of the leaf associated with the `nullifier`.
+    ///
+    /// Conceptually, such a witness is a Merkle path to the leaf, as well as the leaf itself.
+    ///
+    /// This witness is a proof of the current block number of the given nullifier. If that block
+    /// number is zero, it proves that the nullifier is unspent.
+    pub fn open(&self, nullifier: &Nullifier) -> NullifierWitness {
+        NullifierWitness::new(self.smt.open(&nullifier.inner()))
+    }
+
+    /// Returns the block number for the given nullifier or `None` if the nullifier wasn't spent
+    /// yet.
+    pub fn get_block_num(&self, nullifier: &Nullifier) -> Option<BlockNumber> {
+        let value = self.smt.get_value(&nullifier.inner());
+        if value == Smt::EMPTY_VALUE {
+            return None;
+        }
+
+        Some(Self::leaf_value_to_block_num(value))
+    }
+
+    /// Computes mutations for the nullifier SMT.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - a nullifier in the provided iterator was already spent.
+    pub fn compute_mutations(
+        &self,
+        nullifiers: impl IntoIterator<Item = (Nullifier, BlockNumber)>,
+    ) -> Result<NullifierMutationSet, NullifierTreeError> {
+        let mutation_set =
+            self.smt.compute_mutations(nullifiers.into_iter().map(|(nullifier, block_num)| {
+                (nullifier.inner(), Self::block_num_to_leaf_value(block_num))
+            }));
+
+        for nullifier in mutation_set.new_pairs().keys() {
+            let nullifier = Nullifier::from(*nullifier);
+
+            if self.get_block_num(&nullifier).is_some() {
+                return Err(NullifierTreeError::NullifierAlreadySpent(nullifier));
+            }
+        }
+
+        Ok(NullifierMutationSet::new(mutation_set))
+    }
+
+    // PUBLIC MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Marks the given nullifier as spent at the given block number.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - the nullifier was already spent.
```

Comment:
Should we clarify here that this only applies if two nullifiers were spent in different blocks?